Remove dependency in dockerd on libseccomp

cpuguy83 opened the pull request:
This was just using libseccomp to get the right arch, but we can use
GOARCH to get this.
The nativeToSeccomp map needed to be adjusted a bit for mipsle vs mipsel
since that's go how refers to it. Also added some other arches to it.

Closes #41299

cpuguy83 committed:
Remove dependency in dockerd on libseccomp

This was just using libseccomp to get the right arch, but we can use
GOARCH to get this.
The nativeToSeccomp map needed to be adjusted a bit for mipsle vs mipsel
since that's go how refers to it. Also added some other arches to it.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

AkihiroSuda commented on profiles/seccomp/seccomp.go:
```diff
@@ -6,11 +6,11 @@ import (
 	"github.com/docker/docker/api/types"
 	"github.com/docker/docker/pkg/parsers/kernel"
 	specs "github.com/opencontainers/runtime-spec/specs-go"
-	libseccomp "github.com/seccomp/libseccomp-golang"
```
Should be removed from vendor?

AkihiroSuda replied:
Also Dockerfile should no longer need aptgetting libseccomp

cpuguy83 replied:
Still need libseccomp to build runc at least.

cpuguy83 replied:
I may push some images for runc and containerd since these rarely change and just consume those in our Dockerfile...

thaJeztah commented:
Build is failing;

```
[2020-08-27T23:32:31.365Z] === Errors
[2020-08-27T23:32:31.365Z] profiles/seccomp/seccomp.go:102:37: undefined: arch
[2020-08-27T23:32:31.365Z] profiles/seccomp/seccomp.go:121:38: undefined: arch
[2020-08-27T23:32:31.365Z] profiles/seccomp/seccomp.go:102: undefined: arch
[2020-08-27T23:32:31.365Z] profiles/seccomp/seccomp.go:121: undefined: arch
```

cpuguy83 committed:
Remove libseccomp-golang from vendor

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
Updated to remove `seccomp` build tag and instead check at runtime if seccomp is supported.
Also removes vendor and moves libseccomp-dev install to just run `runc` build stage.

thaJeztah commented on profiles/seccomp/default_linux.go:
```diff
@@ -42,6 +40,9 @@ func arches() []types.Architecture {
 
 // DefaultProfile defines the allowed syscalls for the default seccomp profile.
 func DefaultProfile() *types.Seccomp {
+	if !isEnabled() {
```
wondering; what would happen if we didn't even check, and just passed this down to containerd/runc (having them deal with wether or not to do something with it)?

cpuguy83 replied:
Looks like it would error out: https://github.com/opencontainers/runc/blob/2c632d1a2de0192c3f18a2542ccb6f30a8719b1f/libcontainer/seccomp/seccomp_unsupported.go#L16

cpuguy83 replied:
Sorry that's different... when runc is not compiled with seccomp support.

cpuguy83 replied:
Most likely what will happen here is the kernel will throw an error.

thaJeztah replied:
That's if runc applies it; I wonder if runc should do the check, and skip that part of the spec if it's not supported 🤔

Basically; if could have multiple OCI runtimes configured, and the docker daemon doesn't know what they support (hope you follow my train of thought)

cpuguy83 commented:
Hold off on merging... realized there's an issue here where some files are still depending on the 'seccomp' build tag.

cpuguy83 committed:
Memoize seccomp value for SysInfo

As it turns out, we call this function every time someone calls `docker
info`, every time a contianer is created, and every time a container is
started.
Certainly this should be refactored as a whole, but for now, memoize the
seccomp value.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
I pulled back on those buildtag changes just because there's a ton of places where it's used.
I can deal with this in a follow up since it's really a different change.

tianon reviewed (approved):
LGTM (although test failure looks weird)

thaJeztah commented:
@cpuguy83 tests are still failing, so looks related

cpuguy83 commented:
Old profile -> New profile... 🤔 
```diff
>                                 "arch_prctl"
>                         ],
>                         "action": "SCMP_ACT_ALLOW"
>                 },
>                 {
>                         "names": [
>                                 "modify_ldt"
>                         ],
>                         "action": "SCMP_ACT_ALLOW"
>                 },
>                 {
>                         "names": [
```

cpuguy83 commented:
Should be fixed now.
Had to add a mapping from goarch -> libseccomp string -> docker arch type

thaJeztah commented on profiles/seccomp/seccomp_linux.go:
```diff
@@ -79,17 +97,17 @@ func setupSeccomp(config *types.Seccomp, rs *specs.Spec) (*specs.LinuxSeccomp, e
+
+	if len(config.ArchMap) != 0 && archExists {
 		for _, a := range config.ArchMap {
-			seccompArch, ok := nativeToSeccomp[arch]
```
nice catch that this was inside the loop 👍

thaJeztah commented on profiles/seccomp/seccomp_linux.go:
```diff
@@ -29,16 +27,42 @@ func LoadProfile(body string, rs *specs.Spec) (*specs.LinuxSeccomp, error) {
 
+// libseccomp string => seccomp arch
 var nativeToSeccomp = map[string]types.Arch{
+	"x86":         types.ArchX86,
```
couldn't we change this to `goarchToSeccomp` ?

so change this to "386" ?

thaJeztah replied:
ah, nevermind; looks like `arch` is used further down (didn't show in the diff)

thaJeztah replied:
Guess we could still do it, but we would need both mapping (but it would change to GOARCH -> native, and GOARCH -> seccomp)

cpuguy83 replied:
Yeah this was pretty much why it was failing. Didn't realize we were using the libseccomp arch string.

thaJeztah replied:
Do you think it will be cleaner to have `GOARCH -> nativeArch` and `GOARCh -> seccompArch`, so that they can be used independently?

```patch
diff --git a/profiles/seccomp/seccomp_linux.go b/profiles/seccomp/seccomp_linux.go
index 7847085a0a..ac6441d6c3 100644
--- a/profiles/seccomp/seccomp_linux.go
+++ b/profiles/seccomp/seccomp_linux.go
@@ -27,16 +27,16 @@ func LoadProfile(body string, rs *specs.Spec) (*specs.LinuxSeccomp, error) {
        return setupSeccomp(&config, rs)
 }

-// libseccomp string => seccomp arch
-var nativeToSeccomp = map[string]types.Arch{
-       "x86":         types.ArchX86,
+// GOARCH => seccomp arch
+var goToSeccomp = map[string]types.Arch{
+       "386":         types.ArchX86,
        "amd64":       types.ArchX86_64,
        "arm":         types.ArchARM,
        "arm64":       types.ArchAARCH64,
        "mips64":      types.ArchMIPS64,
-       "mips64n32":   types.ArchMIPS64N32,
-       "mipsel64":    types.ArchMIPSEL64,
-       "mips3l64n32": types.ArchMIPSEL64N32,
+       "mips64p32":   types.ArchMIPS64N32,
+       "mips64le":    types.ArchMIPSEL64,
+       "mips64p32le": types.ArchMIPSEL64N32,
        "mipsle":      types.ArchMIPSEL,
        "ppc":         types.ArchPPC,
        "ppc64":       types.ArchPPC64,
@@ -97,8 +97,7 @@ func setupSeccomp(config *types.Seccomp, rs *specs.Spec) (*specs.LinuxSeccomp, e
                }
        }

-       arch := goToNative[runtime.GOARCH]
-       seccompArch, archExists := nativeToSeccomp[arch]
+       seccompArch, archExists := goToSeccomp[runtime.GOARCH]

        if len(config.ArchMap) != 0 && archExists {
                for _, a := range config.ArchMap {
@@ -115,6 +114,8 @@ func setupSeccomp(config *types.Seccomp, rs *specs.Spec) (*specs.LinuxSeccomp, e
        newConfig.DefaultAction = specs.LinuxSeccompAction(config.DefaultAction)

 Loop:
+       arch := goToNative[runtime.GOARCH]
+
        // Loop through all syscall blocks and convert them to libcontainer format after filtering them
        for _, call := range config.Syscalls {
                if len(call.Excludes.Arches) > 0 {
```

cpuguy83 replied:
It's different, I don't think it's better or worse.

thaJeztah replied:
Yeah, it's not a blocker, as non of this is exported. Thought it would be slightly easier to understand, as we wouldn't be mapping twice (GOOS -> native -> seccomp).

I can open a follow-up to discuss

thaJeztah commented on profiles/seccomp/seccomp_linux.go:
```diff
@@ -79,17 +97,17 @@ func setupSeccomp(config *types.Seccomp, rs *specs.Spec) (*specs.LinuxSeccomp, e
 
-	if len(config.ArchMap) != 0 {
+	arch := goToNative[runtime.GOARCH]
+	seccmpArch, archExists := nativeToSeccomp[arch]
```
seccmpArch -> seccompArch?

cpuguy83 replied:
Fixed

thaJeztah reviewed (approved):
LGTM
